staging/src/k8s.io/apiserver/pkg/util/peerproxy/exclusion_filter.go

```diff
@@ -0,0 +1,375 @@
+/*
+Copyright 2025 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package peerproxy
+
+import (
+	"context"
+	"time"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+
+	apidiscoveryv2 "k8s.io/api/apidiscovery/v2"
+)
+
+// GVExtractor is a function that extracts group-versions from an object.
+// It returns a slice of GroupVersions belonging to CRDs or aggregated APIs that
+// should be excluded from peer-discovery to avoid advertising stale CRDs/aggregated APIs
+// in peer-aggregated discovery that were deleted but still appear in a peer's discovery.
+type GVExtractor func(obj interface{}) []schema.GroupVersion
+
+// RegisterCRDInformerHandlers registers event handlers for CRD informer using a custom extractor.
+// The extractor function is responsible for extracting GroupVersions from CRD objects.
+func (h *peerProxyHandler) RegisterCRDInformerHandlers(crdInformer cache.SharedIndexInformer, extractor GVExtractor) error {
+	h.crdInformer = crdInformer
+	h.crdExtractor = extractor
+	_, err := h.crdInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			gvs := extractor(obj)
+			if gvs == nil {
+				return
+			}
+			h.addExcludedGVs(gvs)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			gvs := extractor(newObj)
+			if gvs == nil {
+				return
+			}
+			h.addExcludedGVs(gvs)
+		},
+		DeleteFunc: func(obj interface{}) {
+			// Handle tombstone objects
+			if tombstone, ok := obj.(cache.DeletedFinalStateUnknown); ok {
+				obj = tombstone.Obj
+			}
+			gvs := extractor(obj)
+			if gvs == nil {
+				return
+			}
+			for _, gv := range gvs {
+				h.gvDeletionQueue.Add(gv.String())
+			}
+		},
+	})
+	return err
+}
+
+// RegisterAPIServiceInformerHandlers registers event handlers for APIService informer using a custom extractor.
+// The extractor function is responsible for extracting GroupVersions from APIService objects
+// and determining if they represent aggregated APIs.
+func (h *peerProxyHandler) RegisterAPIServiceInformerHandlers(apiServiceInformer cache.SharedIndexInformer, extractor GVExtractor) error {
+	h.apiServiceInformer = apiServiceInformer
+	h.apiServiceExtractor = extractor
+	_, err := h.apiServiceInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			gvs := extractor(obj)
+			if gvs == nil {
+				return
+			}
+			h.addExcludedGVs(gvs)
```

Comment:
I originally didn't comment that we need a queue here because the operation looked really simple and fast. On a second look, `addExcludedGVs` requires holding a lock. it's possible for `markGVForDeletionIfUsed` to hold the lock and call `isGVUsed`. That operation calls a LIST on both CRDs and APIServices. `addExcludedGVs` would block until the list check is completed. That isn't ideal in eventhandlers, would it be doable to refactor this to use a queue?